v2/cmd/asoctl/pkg/importing/resource_importer.go

```diff
@@ -186,35 +182,35 @@ func (ri *ResourceImporter) importWorker(
 	ctx context.Context,
 	pending <-chan ImportableResource,
 	completed chan<- ImportResourceResult,
-	progress chan<- progressDelta,
+	progress Progress,
 ) {
 	for rsrc := range pending {
 		if ctx.Err() != nil {
-			// If we're aborting, just discard everything until it's all gone
-			progress <- progressDelta{total: -1}
+			// If we're aborting, just remove everything from the queue until it's all gone
+			progress.AddPending(-1)
 			continue
 		}
 
 		// We have a resource to import
-		ri.log.V(1).Info("Importing", "resource", rsrc.Id())
+		ri.log.V(1).Info("Importing", "resource", rsrc.ID())
 		result := ri.ImportResource(ctx, rsrc, progress)
 		completed <- result
 	}
 }
 
 func (ri *ResourceImporter) collateResults(
-	completed <-chan ImportResourceResult,
-	candidates chan<- ImportableResource,
-	progress chan<- progressDelta,
+	completed <-chan ImportResourceResult, // completed imports for us to collate
+	candidates chan<- ImportableResource, // additional candidates for importing
+	progress Progress, // progress tracking
+	publish chan<- *resourceImportReport, // publishing our final summary
 ) {
 	report := newResourceImportReport()
-
 	for importResult := range completed {
 		rsrc := importResult.resource
 		gk := rsrc.GroupKind()
 
-		// Enqueue any child resources we found
-		progress <- progressDelta{total: len(importResult.pending)}
+		// Enqueue any child resources we found; these will be marked as completed when we collate their results
+		progress.AddPending(len(importResult.pending))
 		for _, p := range importResult.pending {
 			candidates <- p
 		}
```

Comment:
It sorta feels like in the programmatic context, if `report.AddFailedImport` is called (because of a failed import) that we should end up returning an `err`?

What do you think?

Reply:
The report collates and deals with the errors, so they're not lost. This method is triggered as a goroutine, so it can't return anything; the result is "returned" via the `publish` channel.

Reply:
I understand that, my point is that the `Import` method (which is the callers entrypoint to all of this) gives them the ability to:
1. Configure a `Progress` (for progress reporting)
2. Configure the resources to import.

but I don't see how they can consume the errors of the import programmatically. They're not lost (because they're written to log), but as far as I can tell the caller doesn't actually get a stack of errors that they can examine in code?

I'm thinking for scenarios where the user is writing their own tool on top of this SDK - they may want to exit with a nonzero exit code when certain resources (say, storage account) are failed to import, but bypass other failures as expected in their environment. 

It's not _100%_ clear to me they can't already do this with the interface provided (maybe by implementing their own `Progress`?) but it's not amazingly clear from the code. I think we should either:
1. Include those details on the `ResourceImportResult`
2. Return an `err` from `Import` along with `ResourceImportResult`, in keeping with the "// Partial results are returned even in the case of an error."
3. Document how the caller can accomplish the above in the documentation, if it can already be easily accomplished.

The fact that there's parallelism under the hood here shouldn't really matter from the perspective of the `Import` method because that's the wrapper for it all, the fan-out and join happen within that method only so from the callers perspective, they made a call, waited a while and got a result, and I think that result should include the ability to programmatically examine errors.

(I probably put this comment in a confusing spot because it's not clear I was talking about `Import` - I think I was tracing through to see if they could already do this with the existing inputs which is what led me to here where we catalog errors, I realize now that's probably confusing though - hopefully the above clarifies.)

Reply:
Ahh, now I understand your point. Yes, extending `ResourceImportResult` is the best way to go - that currently includes the set of successfully imported resources, and we can add to those details of any errors to allow a consumer to make decisions. I'll do that.